chain-abci/src/app/app_init/validator_state.rs

```diff
@@ -0,0 +1,408 @@
+                        ),
+                    liveness_slash_percent,
+                    PunishmentKind::NonLive,
+                )
+            })
+            .collect()
+    }
+
+    pub fn lookup_address(&self, tm_address: &TendermintValidatorAddress) -> &StakedStateAddress {
+        &self.tendermint_validator_addresses[tm_address]
+    }
+
+    pub fn is_current_validator(&self, tm_address: &TendermintValidatorAddress) -> bool {
+        self.tendermint_validator_addresses.contains_key(tm_address)
+    }
+
+    pub fn is_tracked(&self, validator_address: &TendermintValidatorAddress) -> bool {
+        self.punishment
+            .validator_liveness
+            .contains_key(validator_address)
+    }
+
+    /// init -- called for every initial validator
+    pub fn add_initial_validator(
+        &mut self,
+        address: StakedStateAddress,
+        power: TendermintVotePower,
+        node: CouncilNode,
+        block_signing_window: u16,
+    ) {
+        self.validator_state_helper
+            .validator_voting_power
+            .insert(address, power);
+        self.council_nodes_by_power
+            .insert((power, address), node.clone());
+
+        let tendermint_validator_address = TendermintValidatorAddress::from(&node.consensus_pubkey);
+
+        self.add_validator_for_tracking(
+            tendermint_validator_address,
+            address,
+            block_signing_window,
+        );
+    }
+
+    /// add validator for tracking if it wasn't added before
+    pub fn add_validator_for_tracking(
+        &mut self,
+        validator_address: TendermintValidatorAddress,
+        staking_address: StakedStateAddress,
+        block_signing_window: u16,
+    ) {
+        if !self
+            .punishment
+            .validator_liveness
+            .contains_key(&validator_address)
+        {
+            self.tendermint_validator_addresses
+                .insert(validator_address.clone(), staking_address);
+            self.punishment.validator_liveness.insert(
+                validator_address,
+                LivenessTracker::new(block_signing_window),
+            );
+        }
+    }
+
+    /// remove from tracking liveness
+    pub fn remove_validator_from_tracking(
+        &mut self,
+        tendermint_address: &TendermintValidatorAddress,
+    ) {
+        self.punishment
+            .validator_liveness
+            .remove(tendermint_address);
+    }
+}
+
+/// per block, lookup / helper
+#[derive(PartialEq, Debug, Default, Clone)]
+pub struct ValidatorStateHelper {
+    /// validator voting power (current validator set)
+    pub validator_voting_power: BTreeMap<StakedStateAddress, TendermintVotePower>,
+    /// new validator addresses or whose bonded amount changed in the current block
+    power_changed_in_block: BTreeMap<StakedStateAddress, TendermintVotePower>,
+    /// new nodes proposed in the block
+    new_nodes_in_block: BTreeMap<StakedStateAddress, CouncilNode>,
+}
```

Comment:
I think the nature of fields in `ValidatorStateHelper` are different, why put them together separately:
- `validator_voting_power` is an index structure optimized for quering, whose content should keep consistent with `council_nodes_by_power` all the time.
- `power_changed_in_block` and `new_nodes_in_block` are temparary states no needs to be persisted.

Reply:
It was separated it out as in:
`ValidatorState` -- all metadata that needs to be persisted
`ValidatorStateHelper` -- extra info that's either temporary or can be re-constructed from `ValidatorState`

their content may change, may be more fine-grained (e.g. separate out optimized lookup structures and per-block logic handling ones) etc. this is just for start